### PR TITLE
Feature/console and db logic

### PR DESCRIPTION
**Pull request by alepiaz**
The first chunk of feature for this project. 
The idea is to have a console with autocompleting functions that will allow me to fill a database with information about the manga
More functions will be added in the future

Database structure is still WIP and not defined

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+/.vscode
+/db
+__pycache__/
```
Comment (Helias) [inline .gitignore]: puoi usare il template di .gitignore per python che fornisce github

File: README.md [hunk 2]
```diff
@@ -1,3 +1,43 @@
...
+- Gracefully handles RuntimeErrors.
+
+## Installation
+
+1. Clone the repository:
+    ```
+    git clone https://github.com/alepiaz/datapiece.git
+    ```
+2. Navigate to the project directory:
+    ```
```
Comment (Helias) [inline README.md]: usa ```bash in tutti

File: scripts/commands.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+        Args:
+            handler (DBQueryHandler): An instance of DBQueryHandler to execute the queries.
+            config (dict): A dictionary containing config information for commands
+        """
+        self.handler = handler
+        self.config = config
+        self.exclude_list = get_key_list(config, "exclude_list")
+
+    def get_command_names(self):
```
Comment (Helias) [inline scripts/commands.py]: return type `List[something]` or `list[something]` depends if you use 3.9 or 3.10+

File: scripts/commands.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+class Commands:
+    """
+    A class for executing database commands.
+
+    Attributes:
+        handler (DBQueryHandler): Executes the queries.
+    """
+
+    def __init__(self, handler: DBQueryHandler, config: dict):
```
Comment (Helias) [inline scripts/commands.py]: `dict` is ok as type, but optionally you could define a better type that shows also the properties of the dict

File: scripts/commands.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        """
+        Constructs all the necessary attributes for the Commands object.
+
+        Args:
+            handler (DBQueryHandler): An instance of DBQueryHandler to execute the queries.
+            config (dict): A dictionary containing config information for commands
+        """
+        self.handler = handler
+        self.config = config
+        self.exclude_list = get_key_list(config, "exclude_list")
```
Comment (Helias) [inline scripts/commands.py]: not sure if this function is really necessary, you could directly write `config.get("exclude_list", [])`, what are the benefits of making a function?

File: scripts/commands.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+    def get_command_names(self):
+        return [
+            func
+            for func in dir(self)
+            if callable(getattr(self, func)) and not func in self.exclude_list
+        ]
```
Comment (Helias) [inline scripts/commands.py]: not sure if there is a way to make it more readable, the readability is borderline, it's ok for python programmer but if there is a better way to make it more readable I invite you to do it

File: scripts/commands.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+    def start_volume(self, volume_number: int):
```
Comment (Helias) [inline scripts/commands.py]: return type

File: scripts/commands.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        query = f"INSERT INTO Volumes (VolumeNumber) VALUES ({volume_number})"
```
Comment (Helias) [inline scripts/commands.py]:
```suggestion
        query = f"INSERT INTO `Volumes` (`VolumeNumber`) VALUES ({volume_number})"
```

File: scripts/console.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        readline.set_completer(self.completer)
+
+        print('Welcome to the SQL Console. Type "exit" to quit.')
+
+        while True:
+            try:
+                command = readline.readline(">>> ")
+                if command.lower().strip() == "exit":
+                    break
+                else:
```
Comment (Helias) [inline scripts/console.py]: the `else` is not necessary here

File: scripts/console.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        Parameters:
+            handler (DBQueryHandler): An instance of DBQueryHandler for handling database queries.
+            config (dict): A configuration dictionary.
+        """
+        self.handler = handler
+        self.config = config
+        self.commands_instance = Commands(handler, get_key_dict(config, "commands"))
+        self.commands = self.commands_instance.get_command_names()
+
+    def start(self):
```
Comment (Helias) [inline scripts/console.py]: missing return type

File: scripts/console.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+        Returns:
+            str: A completion option that starts with the input text,
+                or None if no more options are available.
+        """
+        options = [i for i in self.commands if i.startswith(text)]
+        if state < len(options):
+            return options[state]
+        else:
+            return None
```
Comment (Helias) [inline scripts/console.py]: the return None or the else are not necessary
you could write directly `return` or you can even omit it and put nothing

File: scripts/console.py [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+            except KeyboardInterrupt:
+                # Handle Ctrl+C
+                continue
+            except EOFError:
+                # Handle Ctrl+D / EOF
+                break
+
+        self.handler.conn.close()
+
+    def completer(self, text: str, state: int):
```
Comment (Helias) [inline scripts/console.py]: missing return type `-> str`

File: README.md (hunk 2 above)
Comment (alepiaz) [reply README.md]: Tutti cosa?

File: scripts/db_query_handler.py [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+        self.create_database_if_needed()
+
+    def handle_database_deletion(self):
+        """
+        Deletes the existing database file if needed and if the file is readable.
+        """
+        if self.need_to_delete() and file_exists_and_is_readable(self.db_path):
+            os.remove(self.db_path)
+
+    def need_to_delete(self):
```
Comment (Helias) [inline scripts/db_query_handler.py]: is_need_to_delete, use prefix is_ or has_ if you have a function that return a boolean
missing return type

File: scripts/commands.py (hunk 5 above)
Comment (alepiaz) [reply scripts/commands.py]: It takes one argument less

File: scripts/db_query_handler.py [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+        """
+        Connects to the SQLite database.
+        """
+        if file_exists_and_dir_is_writable(self.db_path) or not path_exists(
+            self.db_path
+        ):
+            self.conn = sqlite3.connect(self.db_path)
+            self.cursor = self.conn.cursor()
+
+    def create_database_if_needed(self):
```
Comment (Helias) [inline scripts/db_query_handler.py]: I would say if_not_exists to be coherent with SQL commands

File: scripts/db_query_handler.py [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+        """
+        Initializes the database from the schema file.
+
+        Parameters:
+            schema_file (str): Path to the schema file.
+        """
+        sql_commands = self.load_commands_from_schema(schema_file)
+        self.execute_sql_commands_list(sql_commands)
+
+    def load_commands_from_schema(self, schema_file: str) -> list:
```
Comment (Helias) [inline scripts/db_query_handler.py]: -> list of what? put the type

File: scripts/db_query_handler.py [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+        """
+
+        try:
+            with open(schema_file, "r") as f:
+                return f.read().split(";")
+        except FileNotFoundError:
+            print(f"Schema file {schema_file} does not exist.")
+            return []
+
+    def execute_sql_commands_list(self, sql_commands: list):
```
Comment (Helias) [inline scripts/db_query_handler.py]: list[str] I guess

File: scripts/db_query_handler.py [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+
+    def execute_query(self, query: str):
+        """
+        Executes the given SQL query and commits the changes.
+
+        Parameters:
+            query (str): SQL query.
+        """
+        self.cursor.execute(query)
+        self.conn.commit()
```
Comment (Helias) [inline scripts/db_query_handler.py]: in all functions there is no return type hint

File: scripts/utils/os.py [hunk 18]
```diff
@@ -0,0 +1,66 @@
+"""
+A module for checking file and directory permissions.
+"""
+
+import os
+
+
+def path_exists(path):
```
Comment (Helias) [inline scripts/utils/os.py]: is_path_exist

File: scripts/utils/os.py [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+        bool: True if path exists, False otherwise.
+    """
+    try:
+        return os.path.exists(path)
+    except OSError as e:
+        print(f"An OSError occurred while checking if the file exists: {e}")
+        return False
+
+
+def path_is_readable(path):
```
Comment (Helias) [inline scripts/utils/os.py]: is_path_readable

File: scripts/utils/os.py [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+    """
+    return path_exists(file_path) and path_is_writable(file_path)
+
+
+def file_exists_and_dir_is_writable(file_path):
+    """
+    Checks if a file exists and its directory is writable.
+    """
+    dir_path = os.path.dirname(file_path)
+    return path_exists(file_path) and path_is_writable(dir_path)
```
Comment (Helias) [inline scripts/utils/os.py]: add the return type hint, use is_ or has_ for as prefix for functions with boolean as return type

File: README.md (hunk 2 above)
Comment (Helias) [reply README.md]:
```bash
git clone ...
```

```bash
cd datapiece
```

ecc..

File: scripts/commands.py (hunk 4 above)
Comment (alepiaz) [reply scripts/commands.py]: Currently the config is empty for commands section

File: scripts/db_query_handler.py [hunk 21]
```diff
@@ -0,0 +1,158 @@
...
+        self._initialize_database_from_schema(schema_file)
+
+    def _check_tables(self) -> None:
+        """
+        Checks if the necessary tables exist in the database.
+
+        Returns:
+            bool: True if the necessary tables exist, False otherwise.
+        """
+        self.cursor.execute("SELECT name FROM sqlite_master WHERE type='table'")
```
Comment (Helias) [inline scripts/db_query_handler.py]:
```suggestion
        self.cursor.execute("SELECT `name` FROM `sqlite_master` WHERE `type`='table'")
```

File: scripts/db_query_handler.py [hunk 22]
```diff
@@ -0,0 +1,158 @@
...
+        """
+        Creates the database schema if necessary.
+
+        Args:
+            schema_file (str): The path to the schema file containing
+                the database schema definition.
+        """
+        self._initialize_database_from_schema(schema_file)
+
+    def _check_tables(self) -> None:
```
Comment (Helias) [inline scripts/db_query_handler.py]: if_table_exists

is more meaningful than "check_tables"

Comment (alepiaz) [reply scripts/db_query_handler.py]: This goes against the practice of having verbs as function names

File: .github/workflows/main.yml [hunk 23]
```diff
@@ -18,7 +18,7 @@ jobs:
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
-        os: [windows-latest, ubuntu-latest, macos-latest]
+        os: [windows-latest]
```
Comment (Helias) [inline .github/workflows/main.yml:21]: I would add all of them

File: .github/workflows/main.yml [hunk 24]
```diff
@@ -18,7 +18,7 @@ jobs:
...
         python-version: [3.12.0]
```
Comment (Helias) [inline .github/workflows/main.yml]: 3.9 and a 3.10+ python version would be good for back-compatibility

File: .github/workflows/main.yml [hunk 25]
```diff
@@ -37,25 +43,27 @@ jobs:
     - name: Lint with pylint
       run: |
         pip install pylint
-        pylint setup.py tests
-        
+        pylint setup.py scripts tests
```
Comment (Helias) [inline .github/workflows/main.yml]: `pylint $(git ls-files '*.py')`

File: .github/workflows/main.yml [hunk 26]
```diff
@@ -37,25 +43,27 @@ jobs:
...
         pip install flake8
-        flake8 setup.py tests
+        flake8 setup.py scripts tests
+      
         
     - name: Type check with mypy
       run: |
         pip install mypy
-        mypy setup.py tests
+        mypy setup.py scripts tests
```
Comment (Helias) [inline .github/workflows/main.yml]: `mypy $(git ls-files '*.py')`

and/or...
`mypy $(git ls-files '*.py') --check-untyped-defs --ignore-missing-imports`

File: .github/workflows/main.yml [hunk 27]
```diff
@@ -29,6 +29,12 @@ jobs:
...
 
+    - name: Setup Working Directory
+      run: |
+        echo "Current Working Directory is $(pwd)"
+        echo "Adding PWD to Python path"
+        echo "env:PYTHONPATH=$env:PYTHONPATH;$(pwd)" | Out-File -Append -FilePath $env:GITHUB_ENV -Encoding utf8
+
     - name: Install dependencies
       run: |
         python -m pip install --upgrade pip
```
Comment (Helias) [inline .github/workflows/main.yml:40]: below you could put
```
          if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
          if [ -f requirements_dev.txt ]; then pip install -r requirements_dev.txt; fi
```

to install requirements.txt and requirements_dev.txt, and, remove below `pip install pylint flake8 mypy pytest`

File: .github/workflows/main.yml (hunk 24 above)
Comment (Helias) [reply .github/workflows/main.yml]: note: list[str] is not compatible with 3.9, so probably you can't do that

File: scripts/db_query_handler.py [hunk 28]
```diff
@@ -0,0 +1,145 @@
...
+
+import os
+import sqlite3
+
+from scripts.utils.config import get_key_str
+from scripts.utils.files import (is_path_existent, is_readable_existing_file,
+                                 is_writeable_file_directory)
+
+
+class DBQueryHandler:  # pylint: disable=too-few-public-methods
```
Comment (Helias) [inline scripts/db_query_handler.py]: I hate this pylint warning and I always disable it
you could use `.pylintrc` to disable it for all files from conf

File: tests/integration_tests/test_setup_integration.py [hunk 29]
```diff
@@ -0,0 +1,54 @@
...
+import sqlite3
+import unittest
+
+
+class TestSetupIntegration(unittest.TestCase):
+    """
+    This class contains integration tests for the setup
+    """
+
+    def setUp(self) -> None:
```
Comment (Helias) [inline tests/integration_tests/test_setup_integration.py]: what about using snake_case?

File: .github/workflows/main.yml (hunk 23 above)
Comment (alepiaz) [reply .github/workflows/main.yml:21]: I will add them when the project is at a more mature stage, right now the project it's using `pyreadline3` which is a lib available only on Windows based systems.

File: .github/workflows/main.yml (hunk 27 above)
Comment (alepiaz) [reply .github/workflows/main.yml:40]: Good idea

File: tests/integration_tests/test_setup_integration.py (hunk 29 above)
Comment (alepiaz) [reply tests/integration_tests/test_setup_integration.py]: It's a bulit-in of `unittest`

File: .github/workflows/main.yml (hunk 23 above)
Comment (Helias) [reply .github/workflows/main.yml:21]: you are doing something wrong then

**Pull request merged**